src/theory/booleans/theory_bool_rewriter.cpp

```diff
@@ -32,6 +34,236 @@ TheoryBoolRewriter::TheoryBoolRewriter(NodeManager* nm) : TheoryRewriter(nm)
+}
+
+bool TheoryBoolRewriter::addNnfNormChild(std::vector<Node>& children,
+                                         Node c,
+                                         Kind k,
+                                         std::map<Node, bool>& lit_pol,
+                                         bool& childrenChanged)
+{
+  if (k == Kind::OR || k == Kind::AND)
+  {
+    Node lit = c.getKind() == Kind::NOT ? c[0] : c;
+    bool pol = c.getKind() != Kind::NOT;
+    std::map<Node, bool>::iterator it = lit_pol.find(lit);
+    if (it == lit_pol.end())
+    {
+      lit_pol[lit] = pol;
+      children.push_back(c);
+    }
+    else
+    {
+      childrenChanged = true;
+      if (it->second != pol)
+      {
+        return false;
+      }
+    }
+  }
+  else
+  {
+    children.push_back(c);
+  }
+  return true;
+}
+
+Node TheoryBoolRewriter::computeNnfNorm(NodeManager* nm,
+                                        const Node& n,
+                                        TConvProofGenerator* pg)
+{
+  // at pre-order traversal, we store preKind and preChildren, which
+  // determine the Kind and the children for the node to reconstruct.
+  std::unordered_map<TNode, Kind> preKind;
+  std::unordered_map<TNode, std::vector<Node>> preChildren;
+  std::unordered_map<TNode, Node> visited;
+  std::unordered_map<TNode, Node>::iterator it;
+  std::vector<TNode> visit;
+  TNode cur;
+  visit.push_back(n);
+  do
+  {
+    cur = visit.back();
+    visit.pop_back();
+    it = visited.find(cur);
+    if (it == visited.end())
+    {
+      Kind k = cur.getKind();
+      bool negAllCh = false;
+      bool negCh1 = false;
+      // the new formula we should traverse
+      TNode ncur = cur;
+      if (k == Kind::IMPLIES)
+      {
+        k = Kind::OR;
+        negCh1 = true;
+      }
+      else if (k == Kind::XOR)
+      {
+        k = Kind::EQUAL;
+        negCh1 = true;
+      }
+      else if (k == Kind::NOT)
+      {
+        // double negation should already be eliminated
+        Assert(cur[0].getKind() != Kind::NOT);
+        if (cur[0].getKind() == Kind::OR || cur[0].getKind() == Kind::IMPLIES)
+        {
+          k = Kind::AND;
+          negAllCh = true;
+          negCh1 = cur[0].getKind() == Kind::IMPLIES;
+        }
+        else if (cur[0].getKind() == Kind::AND)
+        {
+          k = Kind::OR;
+          negAllCh = true;
+        }
+        else if (cur[0].getKind() == Kind::XOR
+                 || (cur[0].getKind() == Kind::EQUAL
+                     && cur[0][0].getType().isBoolean()))
+        {
+          k = Kind::EQUAL;
+          negCh1 = (cur[0].getKind() == Kind::EQUAL);
+        }
+        else if (cur[0].getKind() == Kind::ITE)
+        {
+          k = cur[0].getKind();
+          negAllCh = true;
+          negCh1 = true;
+        }
+        else
+        {
+          visited[cur] = cur;
+          continue;
+        }
+        ncur = cur[0];
+      }
+      else if ((k != Kind::EQUAL || !cur[0].getType().isBoolean())
```

Comment:
`!cur[0].getType().isBoolean()` was `!body[0].getType().isBoolean()` before. Was it a bug?